src/DependencyInjection/SonataMediaExtension.php

```diff
@@ -276,22 +285,37 @@ public function configureFilesystemAdapter(ContainerBuilder $container, array $c
 
             $arguments = [
                 'region' => $config['filesystem']['s3']['region'],
-                'version' => $config['filesystem']['s3']['version'],
             ];
 
+            if (!$async) {
+                $arguments['version'] = $config['filesystem']['s3']['version'];
+            }
+
             if (isset($config['filesystem']['s3']['endpoint'])) {
                 $arguments['endpoint'] = $config['filesystem']['s3']['endpoint'];
             }
 
             if (isset($config['filesystem']['s3']['secretKey'], $config['filesystem']['s3']['accessKey'])) {
-                $arguments['credentials'] = [
-                    'secret' => $config['filesystem']['s3']['secretKey'],
-                    'key' => $config['filesystem']['s3']['accessKey'],
-                ];
+                if ($async) {
+                    $arguments['accessKeyId'] = $config['filesystem']['s3']['accessKey'];
+                    $arguments['accessKeySecret'] = $config['filesystem']['s3']['secretKey'];
+                } else {
+                    $arguments['credentials'] = [
+                        'secret' => $config['filesystem']['s3']['secretKey'],
+                        'key' => $config['filesystem']['s3']['accessKey'],
+                    ];
+                }
             }
 
-            $container->getDefinition('sonata.media.adapter.service.s3')
-                ->replaceArgument(0, $arguments);
+            if ($async) {
+                $container->getDefinition('sonata.media.adapter.service.s3.async')
```

Comment:
Am I wrong or the definition wont exist if the package is not installed ?

We should add an explicit exception "you need to install foo package to use async"

Reply:
I also just realized that for the full BC we must remain the old s3 client service even when async is turned on, because somebody might rely on that service outside of sonata context. So instead of selecting just one client, I'll configure both if async is turned on.

But I'm still adding errors for both simple and async adapters if the client doesn't exist, it makes sense since both clients are optional and not required by default.

Reply:
I think you don't keep the old s3 client for BC reason because the developer has to set manually the `async` option to true in order to "change" the client service.

But it seems not a bad idea to allow having both clients.